### DISCUSSION TITLE
分享一下个人的懒人更新方案

### DISCUSSION THREAD

**Discussion by GFDGIT** [hooke007/mpv_PlayKit#384] [mpv 通用]
相关讨论：#117、#201 

相信有不少朋友觉得懒人包的更新方式有点繁琐，尤其是在设置了不少个性化选项，还有那种既想追求更新又不想花太多时间折腾的心态。对此，我找到了一种懒人更新方案，只要不是非常巨大的变动，应该可以一劳永逸了。

前置条件：[input-event](https://github.com/natural-harmonia-gropius/input-event)脚本，MPV-lazy的安装目录名为`MPV-lazy`。在此基础我写了一个bat脚本，只要将`mpv-lazy-xxxx.exe`拖到bat上就可以进行更新了。

安装路径的结构如下，主要是多了一个`DATA`文件夹，当然可以随意换成其他名字，但bat和conf要做相应修改：

```
/.../
    └─MPV-lazy
        │  mpv.exe
        │  ...
        ├─DATA
        │  │  custom.conf
        │  │  custom_input.conf
        │  │  custom_profiles.conf
        │  │  升级.bat
        │  ├─scripts
        │  │  inputevent.lua
        │  │  ...
        │  └─_cache
        │      │  saved-props.json
        │      ├─icc
        │      ├─shader
        │      └─watch_later
        ├─installer
        ├─portable_config
        ├─vapoursynth64
        ├─vsgenstubs4
        └─__pycache__
```

其中，需要关注的是`升级.bat`和`custom.conf`文件：

* **升级.bat**
**警告：这个bat会无差别删除它上一级目录下的所有文件，务必不要在工作目录外运行！！！**
```
@echo off
@REM 若bat文件编码为ANSI，注释掉下面一行
chcp 65001

@REM 获取批处理文件所在路径和目录名
cd /d %~dp0
for /f "delims=" %%i in ("%cd%") do set folder=%%~ni

@REM 进入MPV安装目录
cd ..

echo 正在删除旧的程序文件...
del /q *.*
rd /s /q __pycache__ installer portable_config vapoursynth64 vsgenstubs4

echo.
echo 正在解压新安装包...
%1 -y -o%cd%\..

echo.
echo 正在恢复原有数据...
echo.>>.\portable_config/mpv.conf
echo  include = "~~/../%folder%/custom.conf">>.\portable_config\mpv.conf

echo.
echo 升级完成！
pause
```

* **custom.conf**
```
# 数据
watch-later-dir = "~~/../DATA/_cache/watch_later"
gpu-shader-cache-dir = "~~/../DATA/_cache/shader"
icc-cache-dir = "~~/../DATA/_cache/icc"
script-opts-append = save_global_props-cache_dir=~~/../DATA/_cache/

# 加载脚本
scripts-add=~~/../DATA/scripts/auto_sub_fonts_dir.lua
scripts-add=~~/../DATA/scripts/inputevent.lua

# 加载input.conf
script-opts-add = inputevent-configs="~~/input_uosc.conf,~~/../DATA/custom_input.conf"

### MPV设置 ###
include = "~~/../DATA/custom_profiles.conf"
snap-window = yes

### 脚本设置 ###
script-opts-append = thumbfast-hwdec=yes
```

**Comment by GFDGIT** [comment]
这样做的原理是，在`mpv.conf`中include了`custom.conf`，而`custom.conf`中重新设置了会产生数据的目录路径，用`scripts-add`命令加载了`input-event`脚本，再使用`input-event`加载了自定义的`custom_input.conf`。

所以，以后只需要在`custom.conf`和`custom_input.conf`中修改设置，更新时通过bat进行，就可以一步完成，同时不会丢失配置和数据了。

**Reply by hong-tm** [reply]
感觉有意思,钻研一下😗

**Comment by natural-harmonia-gropius** [comment]
补充一点，input-event 加载的额外键位必须有后缀，即使是最基础的单击也需要加 `#@click`

**Reply by GFDGIT** [reply]
是的，而且也不支持向uosc添加菜单项了，因为实际只有`input-event`读取了`custom_input.conf`。

对此，另一种方案是将自定义项写在`input-uosc.conf`后面，每次更新前将修改过的行复制出来，更新完再复制回去，也可以用bat实现，就是写起来比较啰嗦了。

**Comment by zhongfly** [comment]
> 每次更新前将修改过的行复制出来，更新完再复制回去

我觉得不必这样，每次更新时直接每次把lazy包里的input加上你自己的custom_input.conf，合成一个新的文件，然后mpv.conf里引用这个新文件
只不过每次编辑了自己的文件后，得重新合并一次

**Reply by GFDGIT** [reply]
我更新了bat，可以实现input的备份和更新了。

**Comment by GFDGIT** [comment]
更新一下`升级.bat`，不再依赖`input-event`了，并且增加了必要的出错检查。的确显得啰嗦了很多。。。

现在可以将自定义配置写在`input_uosc.conf`中，但要求从最后一行开始写，并以
```
#-#-# custom
```
为起始行，更新时会自动备份此行及以下的所有行，更新完自动恢复。


* **升级.bat**
```
@echo off
@REM 切换代码页，所有文件都需要使用utf-8编码
chcp 65001>nul

@REM 判断升级安装包名称是否符合要求
set filename=%~nx1
set pattern="mpv-lazy-[0-9][0-9][0-9][0-9][0-9][0-9].*exe"
echo "%filename%"|findstr /r %pattern%>nul || goto error1

@REM 获取批处理文件所在路径和数据目录名
cd /d %~dp0
for /f "delims=" %%i in ("%cd%") do set DATA_dir=%%~ni

@REM 判断 MPV-lazy 安装目录名是否符合要求
cd ..
for /f "delims=" %%i in ("%cd%") do set MPV_dir=%%~ni
echo "%MPV_dir%"|findstr /IR "\<mpv-lazy\>">nul || goto error2

@REM 备份 input_uosc.conf 中的自定义部分到 temp.conf
set input_conf=.\portable_config\input_uosc.conf
for /f "tokens=1 delims=:" %%i in ('findstr /n /c:"#-#-# custom" %input_conf%') do ( set /a line=%%i+1 & goto first_match )
:first_match
if DEFINED line (echo 正在备份input.conf设置...) else goto error3
for /f "skip=%line% tokens=1* delims=[]" %%a in ('find /n /v "" %input_conf%') do echo=%%b>>.\%DATA_dir%\temp.conf

:continue
echo 正在删除旧的程序文件...
del /q *.*
rd /s /q __pycache__ installer portable_config vapoursynth64 vsgenstubs4

echo 正在解压新安装包...
%1 -y -o%cd%\..

echo 正在恢复mpv.conf设置...
echo=>>.\portable_config/mpv.conf
echo  include = "~~/../%DATA_dir%/custom.conf">>.\portable_config\mpv.conf

if DEFINED line (
    echo 正在恢复input.conf设置...
    echo=>>.\portable_config\input_uosc.conf
    echo=>>.\portable_config\input_uosc.conf
    type .\%DATA_dir%\temp.conf>>.\portable_config\input_uosc.conf
    del /q .\%DATA_dir%\temp.conf
)

echo 升级完成！ & goto end

:error1
echo 新安装包格式错误! & goto end

:error2
echo 目录名必须为 MPV-lazy（不区分大小写）正在退出！
goto end

:error3
echo input_uosc.conf 中找不到自定义项，请确认!
set /p confirm="输入 y 继续升级，否则退出："
if "%confirm%"=="y" goto continue

:end
echo 按任意键退出 & pause>nul
```

Edit：
1. MPV-lazy安装目录不区分大小写，仍要求完全匹配，否则解压新安装包路径将出错。
2. `msg`命令需要专业版系统，已更换为`echo`。

**Reply by hong-tm** [reply]
mpv.config也是同理吗? 以"#-#-# custom"开头

**Reply by GFDGIT** [reply]
> mpv.config也是同理吗? 以"#-#-# custom"开头

不用，mpv设置是通过include的形式指定的，不需要备份和恢复
```
include = "~~/../%DATA_dir%/custom.conf"
```

**Reply by hong-tm** [reply]
# 跟着步骤在做,拖入exe后terminal界面突然闪烁,没有停留,是有地方错误了吗?

![GIF 2023-11-11 17-23-37](https://github.com/hooke007/MPV_lazy/assets/51999870/3396aeea-5cb1-4eea-b996-089eca4a6630)

## 升级.bat 直接复制粘贴
## custom.conf
```
# 加载脚本
scripts-add=~~/../DATA/scripts/auto_sub_fonts_dir.lua
scripts-add=~~/../DATA/scripts/inputevent.lua

# 加载其他input.conf
script-opts-add = inputevent-configs="input.conf,~~/../DATA/custom_input.conf"

### MPV设置 ###
include = "~~/../DATA/profiles.conf"
snap-window = yes

### 脚本设置 ###
script-opts-append = thumbfast-hwdec=yes

 ########
 # 基础 #
 ########

d3d11-adapter = "Radeon 540"
vulkan-device = "Radeon 540"

 ########
 # 功能 #
 ########

demuxer-max-bytes = 1024MiB
watch-later-dir = "~~/../DATA/_cache/watch_later"
gpu-shader-cache-dir = "~~/../DATA/_cache/shader"
icc-cache-dir = "~~/../DATA/_cache/icc"
script-opts-append = save_global_props-cache_dir=~~/../DATA/_cache/

 ########
 # 视频 #
 ########

 ####################
 # 脚本 滤镜 着色器 #
 ####################

glsl-shaders-append = "~~/shaders/AMD_FSR_EASU_rgb.glsl"
glsl-shaders-append = "~~/shaders/Anime4K_Restore_CNN_S.glsl"
```

## input_osc.conf
```
#-#-# custom
 Tab                 script-binding stats/display-stats-toggle             #! 工具 > 常驻显示统计信息
```

**Reply by GFDGIT** [reply]
你试一下把`升级.bat`第一行的`echo off`改成`echo on`，然后在终端运行（后面的exe路径可以直接拖exe到终端）。
```
升级.bat   mpy-lazy.exe
```
然后看一下报错是什么。

**Reply by hong-tm** [reply]
```
PS C:\Application\mpv-lazy\DATA> .\upgrade.bat 'c:\Users\yi hong\Desktop\mpv-lazy-20231101.exe'

C:\Application\mpv-lazy\DATA>chcp 65001 1>nul

C:\Application\mpv-lazy\DATA>set filename=mpv-lazy-20231101.exe

C:\Application\mpv-lazy\DATA>set pattern="mpv-lazy-[0-9][0-9][0-9][0-9][0-9][0-9].*exe"

C:\Application\mpv-lazy\DATA>echo "mpv-lazy-20231101.exe"  | findstr /r "mpv-lazy-[0-9][0-9][0-9][0-9][0-9][0-9].*exe"  1>nul  || goto error1

C:\Application\mpv-lazy\DATA>cd /d C:\Application\mpv-lazy\DATA\

C:\Application\mpv-lazy\DATA>for /F "delims=" %i in ("C:\Application\mpv-lazy\DATA") do set DATA_dir=%~ni

C:\Application\mpv-lazy\DATA>set DATA_dir=DATA

C:\Application\mpv-lazy\DATA>cd ..

C:\Application\mpv-lazy>for /F "delims=" %i in ("C:\Application\mpv-lazy") do set MPV_dir=%~ni

C:\Application\mpv-lazy>set MPV_dir=mpv-lazy

C:\Application\mpv-lazy>if not "mpv-lazy" == "MPV-lazy" goto error2

C:\Application\mpv-lazy>msg yi hong/w "MPV-lazy的安装目录名不符合要求!"   & exit
'msg' is not recognized as an internal or external command,
operable program or batch file.
```

**Reply by hong-tm** [reply]
MPV-lazy的安装目录名不符合要求!,退出太快看不到,现在终于看到了

**Reply by hong-tm** [reply]
最新的预览版,mpv-lazy-20231101.exe

**Reply by GFDGIT** [reply]
一下子暴露出来3个问题😅

- 安装目录名必须为`MPV-lazy`，这点一开始没有将清楚，而且是区分大小写的。。。
- `msg`命令在win11上不可用。
- `%username%`没有用引号包裹，用户名有空格就会出问题。

你的情况应该是安装目录名没有大写的问题。

**Reply by hooke007** [reply]
话说为什么安装目录用大写的MPV

**Reply by hong-tm** [reply]
解决了:
1. 改了MPV应用目录的名字为```MPV-lazy```
2. ```mpv-lazy-20231101.exe```放进了```DATA```同目录,才拖进```升级.bat```
![image](https://github.com/hooke007/MPV_lazy/assets/51999870/d57b86c9-e5ee-4d3a-9d62-5a62354d76fb)

**Reply by GFDGIT** [reply]
> 话说为什么安装目录用大写的MPV

这个没什么特别的原因，就是我自己一开始用mpv的时候习惯性用了大写，后面就延续了。

这个目录名检查，本意是判断脚本在一个安全的工作目录下执行，避免误删东西。我看看改一下。

**Reply by GFDGIT** [reply]
> 1. 改了MPV应用目录的名字为```MPV-lazy```

名字我改下脚本，之后就不用强制大写了。

> 2. ```mpv-lazy-20231101.exe```放进了```DATA```同目录,才拖进```升级.bat```

不需要在同目录的。

此外，`msg`命令应该是专业版系统才有，也得换掉。

**Reply by hong-tm** [reply]
> > 1. 改了MPV应用目录的名字为`MPV-lazy`
> 
> 名字我改下脚本，之后就不用强制大写了。

可以可以

> > 2. `mpv-lazy-20231101.exe`放进了`DATA`同目录,才拖进`升级.bat`
> 
> 不需要在同目录的。

放在一起不容易出问题

**Reply by hong-tm** [reply]
## 遇到一些问题,interpolation 没有成功关闭,和Anime4K_Restore_CNN_Soft_L.glsl没有成功挂载

```
### MPV设置 ###
include = "~~/../DATA/profiles.conf"
snap-window = yes

### 脚本设置 ###
script-opts-append = thumbfast-hwdec=yes

 ########
 # 基础 #
 ########

fullscreen=yes
d3d11-adapter = "Radeon 540"
vulkan-device = "Radeon 540"

 ########
 # 功能 #
 ########

demuxer-max-bytes = 1024MiB
script-opts-append = save_global_props-cache_dir=~~/../DATA/_cache/

 pause = yes

 ########
 # 视频 #
 ########

video-sync = display-resample
video-sync-max-video-change = 5
interpolation = no

 ####################
 # 脚本 滤镜 着色器 #
 ####################

glsl-shaders-append = "~~/portable_config/shaders/Anime4K_Restore_CNN_Soft_L.glsl"

```

**Reply by hooke007** [reply]
你的vulkan-device的值无效
另外 `~~/` 等价 portable_config 的路径，你写错了

**Reply by hong-tm** [reply]
### 是外接显示器的原因吗?
已经删除,但是好像没有什么变化
```
d3d11-adapter = "Radeon 540"
vulkan-device = "Radeon 540"
```


![image](https://github.com/hooke007/MPV_lazy/assets/51999870/58ce2363-09eb-45d0-9678-f743126a565b)
![image](https://github.com/hooke007/MPV_lazy/assets/51999870/999b5da1-3765-47ca-92e3-3d97c5f6a339)

**Reply by hong-tm** [reply]
> 另外 ~~/ 等价 portable_config 的路径，你写错了

原本是```glsl-shaders-append = "~~/shaders/Anime4K_Restore_CNN_Soft_L.glsls"```但是不知道哪里出问题,所以改了

**Reply by hooke007** [reply]
你可以先看控制台的信息

**Reply by hong-tm** [reply]
### 不会路径的大小号吧? MPV-lazy/mpv-lazy.

目前上面两个问题都还在,我去看看cache
![image](https://github.com/hooke007/MPV_lazy/assets/51999870/da33e052-7ed3-48b3-85e2-961d91147086)

**Reply by GFDGIT** [reply]
> ### 不会路径的大小号吧? MPV-lazy/mpv-lazy.
> 
> 目前上面两个问题都还在,我去看看cache
> ![image](https://github.com/hooke007/MPV_lazy/assets/51999870/da33e052-7ed3-48b3-85e2-961d91147086)
> 

看起来是文件名后面引号不匹配的问题

**Reply by hong-tm** [reply]
> 看起来是文件名后面引号不匹配的问题

多出了```"```我去看看,我可能覆盖了

我的锅,被覆盖了,剩下interpolation关闭的问题,我去排查显卡问题
![image](https://github.com/hooke007/MPV_lazy/assets/51999870/41a78040-53ea-40ec-9af4-e534947424ba)

**Reply by hong-tm** [reply]
我大概看出问题了,```#video-sync = display-resample```在custom.conf里面算是注释了,所以没生效是吧?

![image](https://github.com/hooke007/MPV_lazy/assets/51999870/ba761f4a-ea5c-4572-aa42-09767239f38d)

**Reply by hong-tm** [reply]
目前只能添加不能关闭是吗?🤨我直接在mpv.conf注释是可以的

**Reply by hooke007** [reply]
你这缓存目录写的啥 `script-opts-append = save_global_props-cache_dir=~~/../DATA/_cache/` 非法路径导致脚本崩了

**Reply by hooke007** [reply]
注释掉的东西和没写没区别。mpv不会去识别它
要关掉就显式去关

**Reply by hong-tm** [reply]
> 你这缓存目录写的啥 script-opts-append = save_global_props-cache_dir=~~/../DATA/_cache/

上面一开始抄的时候自带的,没注意他

**Reply by hong-tm** [reply]
> 要关掉就显式去关

~👌,只能进入mpv.conf 关闭~

**Reply by hooke007** [reply]
我的意思是，你应该写成 `video-sync = audio` 而不是 `#video-sync = xxxx`

**Reply by hong-tm** [reply]
> 我的意思是，你应该写成 video-sync = audio 而不是 #video-sync = xxxx

收到,🤣我自己本来想着直接video-sync = no 了,就看到你的回复,回去看了之前的文档
```# 默认值audio（与音频/系统时钟同步）通常兼容性最好但有偶尔的丢帧和重复```

**Reply by GFDGIT** [reply]
> 你这缓存目录写的啥 `script-opts-append = save_global_props-cache_dir=~~/../DATA/_cache/` 非法路径导致脚本崩了

咦，我这边这么写能正常工作啊？

**Reply by hooke007** [reply]
因为他那边没有这个路径，脚本里没有自动创建文件夹的机制

**Reply by GFDGIT** [reply]
> 因为他那边没有这个路径，脚本里没有自动创建文件夹的机制

我将gpu-shader-cache和watch-later的目录也放在这个路径，应该是会自动创建的，除非这两个功能他都关闭了。

**Reply by hong-tm** [reply]
> 我将gpu-shader-cache和watch-later的目录也放在这个路径，应该是会自动创建的，除非这两个功能他都关闭了。

应该是前面某个时候删除了,我想着尽量精简

**Comment by hooke007** [comment]
从上面的例子来看，这个东西对毫无mpv使用基础的来讲还是很折腾的~~
各种意义上想绝对懒人化还得自己很了解mpv

**Reply by hong-tm** [reply]
还需要完善和无脑化🤣

**Reply by GFDGIT** [reply]
额，确实还是需要一点基础的。一方面是脚本不可能写得面面俱到，总会有考虑不周的地方。然后，相对路径写起来也是比较容易出错。

**Reply by letterk** [reply]
> 还需要完善和无脑化🤣

没法无脑化
毕竟portable_config才是懒人包的本体, 里面涉及`scripts`, `shaders`, `vs`
如果出现大的增删, 各种关联全乱, 真要无脑就啥都别改, 一律用默认设置

**Comment by Yunuuuu** [comment]
我目前是配合scoop和git，可以达到完美，更新脚本用 git pull就行，更新mpv用scoop update mpv

**Reply by hong-tm** [reply]
插眼🫠
